### DISCUSSION TITLE
Custom exception handler does not work with 'Exception' class

### DISCUSSION THREAD

**Discussion by ddahan** [fastapi/fastapi#12531] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Description

When building custom exception handlers, everything seems to work as expected, except the "internal_exception_handler handler, supposed to catch "all other exceptions" to display a clean error to the end user.

If replacing `Exception` by `DivisionByZero`, it will work as intended. However, the purpose of this handler is to catch all other exceptions, so I can't list them one by one.

Is it a bug, or is there something wrong in my code?
Thanks.

### Example Code

```python
# exceptions.py

def internal_exception_handler(request: Request, exc: Exception) -> JSONResponse:
    """Handle any other internal exceptions that are not caugh, to raise a 500"""
    logger.exception(f"Internal error occurred: {exc}")
    response = JSONResponse(
        status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
        content={
            "errors": {
                "general": [
                    "An unknown error has occurred. Please contact an administrator."
                ]
            }
        },
    )
    return response

def register_error_handlers(app: FastAPI) -> None:
    # Errors raised manually in endpoint (works)
    app.add_exception_handler(ProjectAPIException, project_api_exception_handler)
    # Errors raised by Pydantic validation error (works)
    app.add_exception_handler(RequestValidationError, validation_exception_handler)
    # Uncaught errors (does not work, it is not called)
    app.add_exception_handler(Exception, internal_exception_handler)


# main.py

app = FastAPI(...)
register_error_handlers(app)

# routes.py

@router.get("/uncaught-exception")
def uncaught_exception():
    result = 1 / 0
    return {"result": result}
```

### Operating System

Linux

### FastAPI Version

0.115.3

### Pydantic Version

2.9.2

### Python Version

3.12.7

**Comment by AumGupta** [comment]
@ddahan your code seems to be absolutely fine.
 
Try running with debug logger once (Make sure your logger is set to display debug messages. You can configure the logging level at the beginning of your application).

```
import logging

logging.basicConfig(level=logging.DEBUG)

``` 

Also check your ASGI Server, ensure you are using an appropriate ASGI server (like `uvicorn` or `daphne`) to run your FastAPI application, as the server can also influence behaviour.

If the problem still exists, please try recreating the issue in a completely isolated environment to ensure no other project settings are affecting the behaviour.

**Comment by ddahan** [comment, marked as answer]
I just found that `app = FastAPI(debug=False)` seems to solve the issue.
In addition, when running unit tests, test client must be instanciated with `TestClient(app, raise_server_exceptions=False)`.
